build.gradle

```diff
@@ -85,51 +218,235 @@ allprojects  {
+            "-Xlint:processing",
+            "-Xlint:removal",
+            "-Xlint:requires-automatic",
+            "-Xlint:requires-transitive-automatic",
+            "-Xlint:try",
+            "-Xlint:varargs",
+            "-Xlint:preview",
+            "-Xlint:static",
+            // "-Werror" // TODO: @vidakovic fix this
+        ]
+        // TODO FINERACT-959 (gradually) enable -Xlint:all (see "javac -help -X")
+
+        options.deprecation = true
+
+        options.compilerArgs +=[
+            "-Xmaxwarns",
+            1500,
+            "-Xmaxerrs",
+            1500
+        ]
+    }
+
+    // Configuration for the spotless plugin
+    // https://github.com/diffplug/spotless/tree/main/plugin-gradle
+    spotless {
+        java {
+            targetExclude '**/build/**', '**/bin/**', '**/out/**'
+            importOrder() //sort imports alphabetically
+            removeUnusedImports()
+            eclipse().configFile "$rootDir/config/fineractdev-formatter.xml"
+            endWithNewline()
+            trimTrailingWhitespace()
+
+            // Enforce style modifier order
+            custom 'Modifier ordering', {
+                def modifierRanking = [
+                    public      : 1,
+                    protected   : 2,
+                    private     : 3,
+                    abstract    : 4,
+                    default     : 5,
+                    static      : 6,
+                    final       : 7,
+                    transient   : 8,
+                    volatile    : 9,
+                    synchronized: 10,
+                    native      : 11,
+                    strictfp    : 12]
+                // Find any instance of multiple modifiers. Lead with a non-word character to avoid
+                // accidental matching against for instance, "an alternative default value"
+                it.replaceAll(/\W(?:public |protected |private |abstract |default |static |final |transient |volatile |synchronized |native |strictfp ){2,}/, {
+                    // Do not replace the leading non-word character.  Identify the modifiers
+                    it.replaceAll(/(?:public |protected |private |abstract |default |static |final |transient |volatile |synchronized |native |strictfp ){2,}/, {
+                        // Sort the modifiers according to the ranking above
+                        it.split().sort({ modifierRanking[it] }).join(' ') + ' '
                     }
                     )
                 }
-            }
-
-            if (project.hasProperty("automatedBuild")) {
-                dependsOn(spotlessCheck)
-            } else {
-                dependsOn(spotlessApply)
+                )
             }
         }
 
+        // TODO: @vidakovic why do we need this again? Spotless is checking anyway
+        // if (project.hasProperty("automatedBuild")) {
+        //     dependsOn(spotlessCheck)
+        // } else {
+        //     dependsOn(spotlessApply)
+        // }
```

Comment:
There was a bit of a debate about this earlier. Initially we just had spotlessCheck which complained about any formatting issues, and then people had to manually run spotlessApply to fix them. There was a fair bit of dissatisfaction about this - and hence this was introduced: if you run build manually it will automatically fix any spotless issues with spotlessApply - but when you run the build in Travis, it just checks and doesn't modify the code. If we want to change that behaviour now that should be a different JIRA & PR up for discussion...

Reply:
Ok, understood. Then I have to move this most likely somewhere else. As is it complains about spotlessApply not being available.

Reply:
Concerning the automatic yaml file generation: the intention is _not_ to regress. Right now unfortunately Gradle is resisting all attempts to order the builds. But sooner or later I'll find a solution for that. So, we are on the same page here.

On that note: there's also a little bit of a chicken and egg problem left there concerning the distribution task. We need to build fineract-provider first which gives us the fineract.yaml file and then we can build fineract-client without the need to hold a hard copy. But as it is right now fineract-provider is also responsible to create the distribution which should include the client library... which would create in this current configuration a circular dependency. Solution: a separate module (fineract-distribution) should be responsible just for creating the distribution file (and signing it etc.). That way we can avoid the circular dependency. Later PR.

Reply:
Found a solutions:

*   automatic spotlessApply works again as before
*   I found a way to express the dependency of fineract-client on fineract-provider; works like a charm; no hard copy of fineract.yaml in fineract-client needed anymore